cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```
@@ -335,39 +365,22 @@ private void publishStopProgram(String stopPayload) {
-                                                                      programUpdateTopic,
-                                                                      programUpdateFetchSize,
-                                                                      lastProgramUpdateMessageId)) {
-        if (iterator.hasNext()) {
-          Message message = iterator.next();
-          Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
-          Map<String, String> properties = notification.getProperties();
-          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
-          String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
-          try {
-            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
-            if (runRecord.getPeerName() != null) {
-              peerProgramUpdates.getOrDefault(runRecord.getPeerName(), new TetheringProgramUpdate())
-                .addUpdateRequest(programRunId, programStatus);
-            }
-          } catch (NotFoundException e) {
-            LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
-          }
-          lastProgramUpdateMessageId = message.getId();
-        }
+      RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
+      if (runRecord.getPeerName() != null) {
+        peerToNotifications.getOrDefault(runRecord.getPeerName(), new ArrayList<>()).add(notification);
       }
-      return peerProgramUpdates;
-    } catch (Exception e) {
-      LOG.error("Exception when fetching program updates. Will retry again during next poll", e);
-      lastProgramUpdateMessageId = lastMessageId;
-      return new HashMap<>();
+    } catch (NotFoundException | IOException e) {
+      LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
```

should let the exception propagate up instead of log and continue. The superclass will retry processing the messages again. Otherwise things like transient DB issues will cause data loss where notifications are skipped.

---

fixed